### PR TITLE
Add option to export actions as a single animation

### PR DESCRIPTION
**Pull request by pjoe**
> **NOTE:** We are in the progress of merging this project into
[glTF-Blender-IO](https://github.com/KhronosGroup/glTF-Blender-IO), a combined importer/exporter.
In the future this repository (glTF-Blender-Exporter) will be archived, and issues and pull
requests closed. We recommend that new issues and pull requests be opened against glTF-Blender-IO, instead.

Submitting PR here for now as it may still be useful for old exporter.

Add option to merge all actions into a single animation when exporting.

This is useful when you have a scene with multiple animated objects, that you need to easily playback as a single animation animation when loaded from glTF.

### CODE REVIEW

**Comment by pjoe** [comment]
Trying to track the new glTF-Blender-IO, but was a bit discouraged from trying to port this PR, when noticed issues about animation export not working yet. Will keep an eye on it though :)

**Comment by donmccurdy** [comment]
@pjoe I realize there are still many open issues with animation in the new project, but as you appear to be using it (thanks again for the PRs!) should we port this PR over? And, do you still think this feature should be included? I'm not quite sure how it would work, for example, when multiple Actions affect a single Armature.

**Comment by ebrayton** [comment]
Can confirm this would be useful. We have animations on several objects in one file and we're interfacing with a tool that makes assumptions about what those separate animations represent. I patched these changes in and the resulting export works flawlessly for our application. Thanks for all your work on these tools.

**Comment by donmccurdy** [comment]
To confirm I understand that correctly – each object has one active Action, and you'd like for the resulting glTF animation to play the active Actions for each object simultaneously?

I believe that's a reasonable addition for the new exporter, but may need require clear naming/documentation to help users understand how to use it. Inactive actions are not associated with anything, and can't be usefully exported that way.

Another option (which should work here, but I haven't tested in the new exporter recently) is to create an NLA Track on each object and give them the same name, which should result in a single merged animation being created. But that is not particularly intuitive. :)

**Comment by ebrayton** [comment]
> To confirm I understand that correctly – each object has one active Action, and you'd like for the resulting glTF animation to play the active Actions for each object simultaneously?

Yes, you understood correctly.

I personally haven't dealt with inactive actions or NLA Tracks, so I don't have any useful comments there. Good to know there's another workaround, though, that although not intuitive, sounds like it'd work for mixed situations where we might want to group some actions into one animation and still be able to have separate animations in the export.

However the implementation works out, documentation on how to deal with combining actions into one exported animation would be very helpful. 👍

**Comment by spiraloid** [comment]
I use separate actions to delineate separate clips for the same armature.  playing them sequentially (not additively) would be my default preference.   that way I could chop up the mater animation by frame number into clips in unity etc. 

i.e. (jump, walk, run erc)
https://youtu.be/RjkXoV84PKw

here’s how multiple clips usually get chopped up from a master animation w fbx. 
https://youtu.be/X7YO6JQE9O8

**Comment by donmccurdy** [comment]
@spiraloid – In this situation I think you would need to create an NLA Track with each Action in sequence and export that. Unfortunately the exporter doesn't support multi-strip NLA Tracks yet, but it's fixable. The problem with any other way of doing it is that inactive Actions are not associated with their object – the exporter cannot tell which clips you want to have in sequence on a given Armature, nor the order in which they should appear.

Also note that glTF supports multiple animations per file – there is no particular need to join them into a single animation only to split them again in Unity.

**Comment by pjoe** [comment]
@donmccurdy I believe the feature and use case is still very valid. For me it is mainly about creating a full scene in Blender with multiple objects and/or armatures that are all animated independently. When then wanting to export the full scene animation and e.g. play back in a web viewer, it quickly becomes unmanageable when each little object has it's own animation in the resulting glTF. Having an option to export the whole scene's animations as a single glTF animation makes this a breeze.

So in short porting this over to glTF-Blender-IO would make a lot of sense IMO. I do seem to recall that I ran into some trouble with this PR though ... maybe when having both object and armature animations in same file.

**Comment by pjoe** [comment]
@spiraloid is that first video Blender 2.49? ... those were the days :)

Anyway I've been using multiple NLA tracks for a character armature with a separate action in each (idle, walk, run, etc) - which exports nicely to multiple animations on same rig in glTF.

**Comment by donmccurdy** [comment]
I'd support having a feature to export all actions (simultaneously, not sequentially) as a single glTF animation. Because glTF has a canonical way of representing multiple animations, I'd prefer not to carry over the workarounds that have been used with other formats.

However, I think it would be better for us to put new features only in glTF-Blender-IO at this point. Closing this PR, but an equivalent PR there would be welcome. Thanks!
